File: examples/python/smoke.py
#!/usr/bin/env python3
# Copyright (c) Andreas Flakstad and Vev contributors
# SPDX-License-Identifier: EPL-2.0

import pathlib
import sys

import vev


def remove_sqlite_files(path: pathlib.Path) -> None:
    for suffix in ("", "-wal", "-shm"):
        try:
            pathlib.Path(f"{path}{suffix}").unlink()
        except FileNotFoundError:
            pass


def main() -> int:
    with vev.open_memory() as conn:
        with conn.transact_report(
            """
            [{:db/id 1 :user/name "Ada" :user/email "ada@example.com"}
             {:db/id 2 :user/name "Grace" :user/email "grace@example.com"}]
            """
        ) as tx:
            tx_value = tx.value()
            print(f"tx: {tx.edn()}")
            if not tx_value.get(":ok") or len(tx_value.get(":tx-data", [])) != 4:
                raise RuntimeError("unexpected typed transaction report")

        conn.transact(
            """
            [[:db/add 90 :db/ident :user/email]
             [:db/add 90 :db/unique :db.unique/identity]]
            """
        )

        collection = conn.query_text(
            """
            [:find ?name
             :in [?email ...]
             :where [?e :user/email ?email]
                    [?e :user/name ?name]]
            """,
            '[["ada@example.com" "grace@example.com"]]',
        )
        print(f"input-collection: {collection}")
        if '"Ada"' not in collection or '"Grace"' not in collection:
            raise RuntimeError("unexpected collection query output")

        conn.transact(
            """
            [[:db/add 100 :db/ident :user/friend]
             [:db/add 100 :db/valueType :db.type/ref]
             [:db/add 1 :user/friend 2]]
            """
        )

        try:
            conn.prepare("[:find ?e :where [?e")
            raise RuntimeError("invalid query unexpectedly prepared")
        except vev.VevError as error:
            if not str(error):
                raise RuntimeError("invalid query did not expose an error") from error

        email_query = conn.prepare(
            """
            [:find ?e ?email
             :in ?needle
             :where [?e :user/email ?email]
                    [(= ?email ?needle)]]
            """
        )
        try:
            with email_query.statement() as stmt:
                rows = stmt.bind("grace@example.com").rows(conn)
                print(f"statement rows: {rows}")
                if rows != [[vev.Entity(2), "grace@example.com"]]:
                    raise RuntimeError("unexpected statement rows")

                streamed = stmt.stream_rows(conn)
                print(f"streamed statement rows: {streamed}")
                if streamed != rows:
                    raise RuntimeError("unexpected streamed statement rows")

                try:
                    stmt.visit(conn, lambda event, row, index, value: False)
                    raise RuntimeError("cancelled visitor unexpectedly succeeded")
                except vev.VevError as error:
                    if "cancelled" not in str(error):
                        raise RuntimeError("unexpected visitor cancellation error") from error

                rows = stmt.bind("ada@example.com").rows(conn)
                print(f"statement rebound rows: {rows}")
                if rows != [[vev.Entity(1), "ada@example.com"]]:
                    raise RuntimeError("unexpected rebound statement rows")

            collection_query = conn.prepare(
                """
                [:find ?name
                 :in [?email ...]
                 :where [?e :user/email ?email]
                        [?e :user/name ?name]]
                """
            )
            try:
                with collection_query.statement() as stmt:
                    rows = stmt.bind(["ada@example.com", "grace@example.com"]).rows(conn)
                    names = sorted(row[0] for row in rows)
                    print(f"statement collection names: {names}")
                    if names != ["Ada", "Grace"]:
                        raise RuntimeError("unexpected collection statement rows")
                with collection_query.statement() as stmt:
                    rows = stmt.bind(vev.TypedCollection("string")).rows(conn)
                    if rows != []:
                        raise RuntimeError("unexpected empty typed collection rows")
            finally:
                collection_query.close()

            all_emails_query = conn.prepare(
                """
                [:find ?email
                 :where [?e :user/email ?email]]
                """
            )
            try:
                with conn.db() as db:
                    columns = db.query_columns(all_emails_query)
                    if columns is None:
                        raise RuntimeError("expected string column batch")
                    emails = sorted(row[0] for row in columns.rows())
                    print(f"column batch emails: {emails}")
                    if columns.kinds != (vev.VEV_COLUMN_STRING,):
                        raise RuntimeError("unexpected column batch kind")
                    if emails != ["ada@example.com", "grace@example.com"]:
                        raise RuntimeError("unexpected column batch rows")

                with all_emails_query.statement() as stmt:
                    columns = stmt.columns(conn)
                    if columns is None:
                        raise RuntimeError("expected live statement column batch")
                    emails = sorted(row[0] for row in columns.rows())
                    if columns.kinds != (vev.VEV_COLUMN_STRING,):
                        raise RuntimeError("unexpected live statement column batch kind")
                    if emails != ["ada@example.com", "grace@example.com"]:
                        raise RuntimeError("unexpected live statement column batch rows")

                    with conn.db() as db:
                        columns = stmt.columns(db)
                        if columns is None:
                            raise RuntimeError("expected snapshot statement column batch")
                        emails = sorted(row[0] for row in columns.rows())
                        if columns.kinds != (vev.VEV_COLUMN_STRING,):
                            raise RuntimeError(
                                "unexpected snapshot statement column batch kind"
                            )
                        if emails != ["ada@example.com", "grace@example.com"]:
                            raise RuntimeError(
                                "unexpected snapshot statement column batch rows"
                            )
            finally:
                all_emails_query.close()

            tuple_query = conn.prepare(
                """
                [:find ?e
                 :in [?name ?email]
                 :where [?e :user/name ?name]
                        [?e :user/email ?email]]
                """
            )
            try:
                with tuple_query.statement() as stmt:
                    rows = stmt.bind(vev.TupleInput(("Ada", "ada@example.com"))).rows(conn)
                    print(f"tuple statement rows: {rows}")
                    if rows != [[vev.Entity(1)]]:
                        raise RuntimeError("unexpected tuple statement rows")
            finally:
                tuple_query.close()

            relation_query = conn.prepare(
                """
                [:find ?name ?label
                 :in [[?email ?label]]
                 :where [?e :user/email ?email]
                        [?e :user/name ?name]]
                """
            )
            try:
                with relation_query.statement() as stmt:
                    rows = stmt.bind(
                        vev.Relation(
                            (
                                ("ada@example.com", "primary"),
                                ("missing@example.com", "missing"),
                            )
                        )
                    ).rows(conn)
                    print(f"relation statement rows: {rows}")
                    if rows != [["Ada", "primary"]]:
                        raise RuntimeError("unexpected relation statement rows")
            finally:
                relation_query.close()

            lookup_query = conn.prepare(
                """
                [:find ?name
                 :in ?person
                 :where [?person :user/name ?name]]
                """
            )
            try:
                with lookup_query.statement() as stmt:
                    rows = stmt.bind(vev.LookupRef(":user/email", "ada@example.com")).rows(conn)
                    print(f"lookup-ref statement rows: {rows}")
                    if rows != [["Ada"]]:
                        raise RuntimeError("unexpected lookup-ref statement rows")
            finally:
                lookup_query.close()

            conn.transact(
                """
                [[:db/add 91 :db/ident :user/code]
                 [:db/add 91 :db/unique :db.unique/identity]
                 [:db/add 1 :user/code 101]
                 [:db/add 2 :user/code 202]]
                """
            )

            lookup_collection_query = conn.prepare(
                """
                [:find ?name
                 :in [?person ...]
                 :where [?person :user/name ?name]]
                """
            )
            try:
                with lookup_collection_query.statement() as stmt:
                    rows = stmt.bind(
                        [
                            vev.LookupRef(":user/email", "ada@example.com"),
                            vev.LookupRef(":user/email", "grace@example.com"),
                        ]
                    ).rows(conn)
                    names = sorted(row[0] for row in rows)
                    print(f"lookup-ref collection statement names: {names}")
                    if names != ["Ada", "Grace"]:
                        raise RuntimeError("unexpected lookup-ref collection statement rows")
                with lookup_collection_query.statement() as stmt:
                    rows = stmt.bind(
                        [
                            vev.LookupRef(":user/code", 101),
                            vev.LookupRef(":user/code", 202),
                        ]
                    ).rows(conn)
                    names = sorted(row[0] for row in rows)
                    if names != ["Ada", "Grace"]:
                        raise RuntimeError("unexpected int lookup-ref collection rows")
            finally:
                lookup_collection_query.close()

            pull_query = conn.prepare(
                """
                [:find (pull ?e [:user/name {:user/friend [:user/name]}])
                 :where [?e :user/name "Ada"]]
                """
            )
            try:
                pulled = pull_query.scalar(conn)
                print(f"pull: {pulled}")
                if (
                    pulled.get(":user/name") != "Ada"
                    or pulled.get(":user/friend", {}).get(":user/name") != "Grace"
                ):
                    raise RuntimeError("unexpected pull result")
            finally:
                pull_query.close()

            with conn.db() as pull_db:
                direct_pull = pull_db.pull(
                    "[:user/name {:user/friend [:user/name]}]", vev.Entity(1)
                )
                print(f"direct pull: {direct_pull}")
                if (
                    direct_pull.get(":user/name") != "Ada"
                    or direct_pull.get(":user/friend", {}).get(":user/name") != "Grace"
                ):
                    raise RuntimeError("unexpected direct pull")

                lookup_pull = pull_db.pull_lookup_ref(
                    "[:user/name]",
                    vev.LookupRef(":user/email", "ada@example.com"),
                )
                print(f"lookup pull: {lookup_pull}")
                if lookup_pull.get(":user/name") != "Ada":
                    raise RuntimeError("unexpected lookup-ref pull")

                many_pull = pull_db.pull_many("[:user/name]", [vev.Entity(1), vev.Entity(2)])
                print(f"pull many: {many_pull}")
                names = sorted(item.get(":user/name") for item in many_pull)
                if names != ["Ada", "Grace"]:
                    raise RuntimeError("unexpected pull-many")

            pull_pattern_query = conn.prepare(
                """
                [:find (pull ?e ?pattern)
                 :in ?pattern ?name
                 :where [?e :user/name ?name]]
                """
            )
            try:
                with pull_pattern_query.statement() as stmt:
                    pulled = stmt.bind(
                        vev.PullPattern("[:user/name {:user/friend [:user/name]}]"),
                        "Ada",
                    ).scalar(conn)
                    print(f"statement pull pattern: {pulled}")
                    if (
                        pulled.get(":user/name") != "Ada"
                        or pulled.get(":user/friend", {}).get(":user/name") != "Grace"
                    ):
                        raise RuntimeError("unexpected statement pull pattern")
            finally:
                pull_pattern_query.close()

            with vev.open_memory() as right_conn:
                right_conn.transact(
                    """
                    [{:db/id 1 :user/name "Ada Right"}
                     {:db/id 2 :user/name "Grace Right"}]
                    """
                )
                with conn.db() as left_db, right_conn.db() as right_db:
                    source_query = conn.prepare(
                        """
                        [:find ?e ?left-name ?right-name
                         :in $left $right [?e ...]
                         :where [$left ?e :user/name ?left-name]
                                [$right ?e :user/name ?right-name]]
                        """,
                        ["$left", "$right"],
                    )
                    try:
                        with source_query.statement() as stmt:
                            rows = stmt.bind(
                                vev.DBSource("$left", left_db),
                                vev.DBSource("$right", right_db),
                                [vev.Entity(1), vev.Entity(2)],
                            ).rows(conn)
                            print(f"statement DB sources: {rows}")
                            first = rows[0]
                            if (
                                len(rows) != 2
                                or first[1] != "Ada"
                                or first[2] != "Ada Right"
                            ):
                                raise RuntimeError("unexpected statement DB sources")
                    finally:
                        source_query.close()

            all_emails = conn.prepare(
                "[:find ?e ?email :where [?e :user/email ?email]]"
            )
            try:
                with conn.db() as snapshot:
                    conn.transact(
                        '[{:db/id 3 :user/name "Alan" :user/email "alan@example.com"}]'
                    )

                    current_rows = len(all_emails.rows(conn))
                    snapshot_rows = len(all_emails.rows(snapshot))
                    print(f"current-db rows: {current_rows}")
                    print(f"snapshot-db rows: {snapshot_rows}")
                    if current_rows != 3 or snapshot_rows != 2:
                        raise RuntimeError("unexpected snapshot row counts")

                    with conn.prepare(
                        '[:find ?e :where [?e :user/name "Barbara"]]'
                    ) as barbara_query, conn.prepare(
                        '[:find ?e :where [?e :user/name "Dorothy"]]'
                    ) as dorothy_query:
                        with snapshot.with_report(
                            '[{:db/id 4 :user/name "Barbara"}]'
                        ) as report:
                            report_value = report.value()
                            print(f"with-db: {report.edn()}")
                            if not report_value.get(":ok"):
                                raise RuntimeError("unexpected with report")
                        with snapshot.db_with(
                            '[{:db/id 4 :user/name "Barbara"}]'
                        ) as next_db:
                            if len(barbara_query.rows(snapshot)) != 0:
                                raise RuntimeError("db-with mutated source DB")
                            if len(barbara_query.rows(next_db)) != 1:
                                raise RuntimeError("db-with missing new fact")
                            with vev.Connection.from_db(next_db) as derived:
                                derived.transact(
                                    '[{:db/id 5 :user/name "Dorothy"}]'
                                )
                                if (
                                    len(barbara_query.rows(derived)) != 1
                                    or len(dorothy_query.rows(derived)) != 1
                                ):
                                    raise RuntimeError(
                                        "conn-from-db did not initialize from DB"
                                    )
            finally:
                all_emails.close()
        finally:
            email_query.close()

    sqlite_path = pathlib.Path("tmp.vev.python.sqlite")
    remove_sqlite_files(sqlite_path)
    try:
        with vev.connect(sqlite_path) as durable:
            if durable.backend() != "sqlite" or durable.path() != str(sqlite_path):
                raise RuntimeError("unexpected durable connection metadata")
            if durable.basis_t() != 0:
                raise RuntimeError("unexpected initial durable basis")
            if durable.tx_count() != 0:
                raise RuntimeError("unexpected initial durable tx count")
            if durable.tx_ids() != []:
                raise RuntimeError("unexpected initial durable tx ids")
            info = durable.info_edn()
            if (
                ":backend :sqlite" not in info
                or ":basis-t 0" not in info
                or ":tx-count 0" not in info
            ):
                raise RuntimeError("unexpected durable connection info")
            with durable.transact_report(
                '[{:db/id 1 :user/name "Durable Ada" :user/email "durable-ada@example.com"}]'
            ) as report:
                if not report.value().get(":ok"):
                    raise RuntimeError("unexpected SQLite transaction report")
            if durable.basis_t() != 1:
                raise RuntimeError("unexpected durable basis after first tx")
            if durable.tx_count() != 1:
                raise RuntimeError("unexpected durable tx count after first tx")
            if durable.tx_ids() != [1]:
                raise RuntimeError("unexpected durable tx ids after first tx")
            with durable.prepare(
                "[:find ?e ?email :where [?e :user/email ?email]]"
            ) as all_emails, durable.db() as db:
                rows = all_emails.rows(db)
                print(f"sqlite-live rows: {rows}")
                if len(rows) != 1:
                    raise RuntimeError("unexpected SQLite live row count")

        with vev.connect(sqlite_path) as durable:
            if durable.basis_t() != 1:
                raise RuntimeError("unexpected reopened durable basis")
            if durable.tx_count() != 1:
                raise RuntimeError("unexpected reopened durable tx count")
            if durable.tx_ids() != [1]:
                raise RuntimeError("unexpected reopened durable tx ids")
            with durable.prepare(
                "[:find ?e ?email :where [?e :user/email ?email]]"
            ) as all_emails, durable.db() as db:
                rows = all_emails.rows(db)
                print(f"sqlite-reopened rows: {rows}")
                if len(rows) != 1:
                    raise RuntimeError("unexpected SQLite reopened row count")
    finally:
        remove_sqlite_files(sqlite_path)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/java/Smoke.java
// Copyright (c) Andreas Flakstad and Vev contributors
// SPDX-License-Identifier: EPL-2.0

package vev;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Arrays;
import java.util.Map;

public final class Smoke {
    private static void deleteSqliteFiles(Path path) throws Exception {
        Files.deleteIfExists(path);
        Files.deleteIfExists(Path.of(path.toString() + "-wal"));
        Files.deleteIfExists(Path.of(path.toString() + "-shm"));
    }

    public static void main(String[] args) throws Throwable {
        if (args.length != 1) {
            throw new IllegalArgumentException("usage: Smoke <path-to-libvev.dylib>");
        }

        Vev vev = Vev.load(Path.of(args[0]));
        try (Vev.Connection conn = vev.createConn()) {
            String tx = conn.transact("""
                [{:db/id 1 :user/name "Ada" :user/email "ada@example.com"}
                 {:db/id 2 :user/name "Grace" :user/email "grace@example.com"}]
                """);
            System.out.println("tx: " + tx);

            String collectionText = conn.queryText("""
                [:find ?name
                 :in [?email ...]
                 :where [?e :user/email ?email]
                        [?e :user/name ?name]]
                """, "[[\"ada@example.com\" \"grace@example.com\"]]");
            System.out.println("input-collection: " + collectionText);
            if (!collectionText.contains("\"Ada\"") || !collectionText.contains("\"Grace\"")) {
                throw new IllegalStateException("unexpected collection query output");
            }

            try (Vev.DB db = conn.db()) {
                List<List<Object>> requestRows = vev.queryRows(Map.of(
                    "query", """
                        [:find ?name
                         :in $ ?email
                         :where [?e :user/email ?email]
                                [?e :user/name ?name]]
                        """,
                    "args", List.of(db, "ada@example.com")));
                System.out.println("query request rows: " + requestRows);
                if (!requestRows.equals(List.of(List.of("Ada")))) {
                    throw new IllegalStateException("unexpected query request rows");
                }

                List<Map<Object, Object>> requestMaps = vev.queryMaps(Map.of(
                    "query", """
                        [:find ?name ?email
                         :keys name email
                         :in $ ?email
                         :where [?e :user/email ?email]
                                [?e :user/name ?name]]
                        """,
                    "args", List.of(db, "ada@example.com")));
                System.out.println("query request maps: " + requestMaps);
                if (!requestMaps.equals(List.of(Map.of(
                        new Vev.Keyword(":name"), "Ada",
                        new Vev.Keyword(":email"), "ada@example.com")))) {
                    throw new IllegalStateException("unexpected query request maps");
                }
            }

            conn.transact("""
                [[:db/add 90 :db/ident :user/email]
                 [:db/add 90 :db/unique :db.unique/identity]]
                """);

            conn.transact("""
                [[:db/add 91 :db/ident :user/status]
                 [:db/add 91 :db/unique :db.unique/identity]
                 [:db/add 92 :db/ident :user/code]
                 [:db/add 92 :db/unique :db.unique/identity]
                 [:db/add 1 :user/status :active]
                 [:db/add 2 :user/status :inactive]
                 [:db/add 1 :user/code 1001]
                 [:db/add 2 :user/code 1002]]
                """);

            conn.transact("""
                [[:db/add 100 :db/ident :user/friend]
                 [:db/add 100 :db/valueType :db.type/ref]
                 [:db/add 1 :user/friend 2]]
                """);

            try (Vev.PreparedQuery emailQuery = vev.prepare("""
                    [:find ?e ?email
                     :in ?needle
                     :where [?e :user/email ?email]
                            [(= ?email ?needle)]]
                    """);
                 Vev.Statement stmt = emailQuery.statement()) {
                String preparedAst = emailQuery.edn();
                if (!preparedAst.contains(":clauses") || !preparedAst.contains(":input-specs")) {
                    throw new IllegalStateException("prepared query AST did not expose parser keys");
                }

                try (Vev.ResultSet result = conn.query(stmt.bindString("grace@example.com"))) {
                    List<List<Object>> rows = result.rows();
                    System.out.println("statement rows: " + rows);
                    if (!rows.equals(List.of(List.of(new Vev.Entity(2), "grace@example.com")))) {
                        throw new IllegalStateException("unexpected statement rows");
                    }
                }

                try (Vev.PreparedQuery collectionQuery = vev.prepare("""
                        [:find ?name
                         :in [?email ...]
                         :where [?e :user/email ?email]
                                [?e :user/name ?name]]
                        """);
                     Vev.Statement collectionStmt = collectionQuery.statement()) {
                    try (Vev.ResultSet result = conn
                            .query(collectionStmt.bindStringCollection("ada@example.com", "grace@example.com"))) {
                        List<List<Object>> collectionRows = result.rows();
                        List<String> names = new ArrayList<>();
                        for (List<Object> row : collectionRows) {
                            names.add((String) row.get(0));
                        }
                        Collections.sort(names);
                        System.out.println("statement collection names: " + names);
                        if (!names.equals(List.of("Ada", "Grace"))) {
                            throw new IllegalStateException("unexpected collection statement rows");
                        }
                    }
                }

                try (Vev.PreparedQuery nameQuery = vev.prepare("""
                        [:find ?name
                         :in ?email
                         :where [?e :user/email ?email]
                                [?e :user/name ?name]]
                        """);
                     Vev.Statement nameStmt = nameQuery.statement();
                     Vev.DB snapshot = conn.db()) {
                    Vev.ColumnResult nameColumns = snapshot.queryColumns(nameStmt.bindString("grace@example.com"));
                    if (nameColumns == null
                            || nameColumns.rowCount() != 1
                            || !Arrays.equals(nameColumns.kinds(), new int[] { Vev.COLUMN_STRING })
                            || !Arrays.equals((String[]) nameColumns.columns()[0], new String[] { "Grace" })) {
                        throw new IllegalStateException("unexpected statement column batch");
                    }
                }

                try (Vev.PreparedQuery pullQuery = vev.prepare("""
                        [:find (pull ?e [:user/name {:user/friend [:user/name]}])
                         :where [?e :user/name "Ada"]]
                        """)) {
                    try (Vev.ResultSet result = conn.query(pullQuery, "[]")) {
                        Vev.MapValue pulled = (Vev.MapValue) result.scalar();
                        System.out.println("pull: " + pulled);
                        Object friend = pulled.get(":user/friend");
                        if (!"Ada".equals(pulled.get(":user/name"))
                            || !(friend instanceof Vev.MapValue friendMap)
                            || !"Grace".equals(friendMap.get(":user/name"))) {
                            throw new IllegalStateException("unexpected pull result");
                        }
                    }
                }

                try (Vev.DB pullDb = conn.db()) {
                    Vev.MapValue directPull = (Vev.MapValue) pullDb.pull(
                        "[:user/name {:user/friend [:user/name]}]",
                        1);
                    System.out.println("direct pull: " + directPull);
                    Object friend = directPull.get(":user/friend");
                    if (!"Ada".equals(directPull.get(":user/name"))
                        || !(friend instanceof Vev.MapValue friendMap)
                        || !"Grace".equals(friendMap.get(":user/name"))) {
                        throw new IllegalStateException("unexpected direct pull");
                    }

                    Vev.MapValue lookupPull = (Vev.MapValue) pullDb.pullLookupRefString(
                        "[:user/name]",
                        ":user/email",
                        "ada@example.com");
                    System.out.println("lookup pull: " + lookupPull);
                    if (!"Ada".equals(lookupPull.get(":user/name"))) {
                        throw new IllegalStateException("unexpected lookup-ref pull");
                    }

                    Vev.MapValue keywordLookupPull = (Vev.MapValue) pullDb.pullLookupRefKeyword(
                        "[:user/name]",
                        ":user/status",
                        ":active");
                    Vev.MapValue intLookupPull = (Vev.MapValue) pullDb.pullLookupRefInt(
                        "[:user/name]",
                        ":user/code",
                        1001);
                    if (!"Ada".equals(keywordLookupPull.get(":user/name"))
                        || !"Ada".equals(intLookupPull.get(":user/name"))) {
                        throw new IllegalStateException("unexpected typed lookup-ref pull");
                    }

                    @SuppressWarnings("unchecked")
                    List<Object> manyPull = (List<Object>) pullDb.pullMany("[:user/name]", 1, 2);
                    List<String> pullNames = new ArrayList<>();
                    for (Object item : manyPull) {
                        pullNames.add((String) ((Vev.MapValue) item).get(":user/name"));
                    }
                    Collections.sort(pullNames);
                    System.out.println("pull many: " + manyPull);
                    if (!pullNames.equals(List.of("Ada", "Grace"))) {
                        throw new IllegalStateException("unexpected pull-many");
                    }
                }

                try (Vev.PreparedQuery pullPatternQuery = vev.prepare("""
                        [:find (pull ?e ?pattern)
                         :in ?pattern ?name
                         :where [?e :user/name ?name]]
                        """);
                     Vev.PreparedPullPattern preparedPattern =
                         vev.preparePullPattern("[:user/name {:user/friend [:user/name]}]");
                     Vev.Statement pullPatternStmt = pullPatternQuery.statement();
                     Vev.ResultSet result = conn.query(
                         pullPatternStmt.bindPullPatternAndString(
                             "[:user/name {:user/friend [:user/name]}]",
                             "Ada"))) {
                    String preparedPatternEdn = preparedPattern.edn();
                    if (!preparedPatternEdn.contains(":pattern")
                        || !preparedPatternEdn.contains(":attr")
                        || !preparedPatternEdn.contains(":nested-count")) {
                        throw new IllegalStateException("unexpected prepared pull pattern EDN: " + preparedPatternEdn);
                    }
                    Vev.MapValue pulled = (Vev.MapValue) result.scalar();
                    System.out.println("statement pull pattern: " + pulled);
                    Object friend = pulled.get(":user/friend");
                    if (!"Ada".equals(pulled.get(":user/name"))
                        || !(friend instanceof Vev.MapValue friendMap)
                        || !"Grace".equals(friendMap.get(":user/name"))) {
                        throw new IllegalStateException("unexpected statement pull pattern");
                    }
                }

                try (Vev.PreparedQuery allEmails = vev.prepare("[:find ?e ?email :where [?e :user/email ?email]]");
                     Vev.DB snapshot = conn.db()) {
                    conn.transact("[{:db/id 3 :user/name \"Alan\" :user/email \"alan@example.com\"}]");
                    try (Vev.ResultSet current = conn.query(allEmails, "[]");
                         Vev.ResultSet old = snapshot.query(allEmails, "[]")) {
                        int currentRows = current.rowCount();
                        int snapshotRows = old.rowCount();
                        System.out.println("current-db rows: " + currentRows);
                        System.out.println("snapshot-db rows: " + snapshotRows);
                        if (currentRows != 3 || snapshotRows != 2) {
                            throw new IllegalStateException("unexpected snapshot row counts");
                        }
                    }

                    try (Vev.ResultSet result = snapshot.query(stmt.bindString("ada@example.com"))) {
                        List<List<Object>> snapshotStmtRows = result.rows();
                        if (!snapshotStmtRows.equals(List.of(List.of(new Vev.Entity(1), "ada@example.com")))) {
                            throw new IllegalStateException("unexpected snapshot statement rows");
                        }
                    }
                }
            }

            Path sqlitePath = Path.of("tmp.vev.java.sqlite");
            deleteSqliteFiles(sqlitePath);
            try {
                try (Vev.DurableConnection durable = vev.connect(sqlitePath)) {
                    if (!"sqlite".equals(durable.backend()) || !sqlitePath.toString().equals(durable.path())) {
                        throw new IllegalStateException("unexpected durable connection metadata");
                    }
                    if (durable.basisT() != 0) {
                        throw new IllegalStateException("unexpected initial durable basis");
                    }
                    if (durable.txCount() != 0) {
                        throw new IllegalStateException("unexpected initial durable tx count");
                    }
                    if (!Arrays.equals(durable.txIds(), new long[]{})) {
                        throw new IllegalStateException("unexpected initial durable tx ids");
                    }
                    String info = durable.infoEdn();
                    if (!info.contains(":backend :sqlite") || !info.contains(":basis-t 0") || !info.contains(":tx-count 0")) {
                        throw new IllegalStateException("unexpected durable connection info");
                    }
                    try (Vev.TxReport report = durable.transactReport("""
                            [{:db/id 1
                              :user/name "Durable Ada"
                              :user/email "durable-ada@example.com"}]
                            """)) {
                        Vev.MapValue reportValue = (Vev.MapValue) report.value();
                        if (!Boolean.TRUE.equals(reportValue.get(":ok"))) {
                            throw new IllegalStateException("unexpected SQLite transaction report");
                        }
                    }
                    if (durable.basisT() != 1) {
                        throw new IllegalStateException("unexpected durable basis after first tx");
                    }
                    if (durable.txCount() != 1) {
                        throw new IllegalStateException("unexpected durable tx count after first tx");
                    }
                    if (!Arrays.equals(durable.txIds(), new long[]{1})) {
                        throw new IllegalStateException("unexpected durable tx ids after first tx");
                    }
                    try (Vev.PreparedQuery durableQuery = vev.prepare("[:find ?e ?email :where [?e :user/email ?email]]");
                         Vev.DB durableDb = durable.db();
                         Vev.ResultSet rows = durableDb.query(durableQuery, "[]")) {
                        System.out.println("sqlite-live rows: " + rows.rowCount());
                        if (rows.rowCount() != 1) {
                            throw new IllegalStateException("unexpected SQLite live row count");
                        }
                    }
                }

                try (Vev.DurableConnection durable = vev.connect(sqlitePath);
                     Vev.PreparedQuery durableQuery = vev.prepare("[:find ?e ?email :where [?e :user/email ?email]]");
                     Vev.DB durableDb = durable.db();
                     Vev.ResultSet rows = durableDb.query(durableQuery, "[]")) {
                    if (durable.basisT() != 1) {
                        throw new IllegalStateException("unexpected reopened durable basis");
                    }
                    if (durable.txCount() != 1) {
                        throw new IllegalStateException("unexpected reopened durable tx count");
                    }
                    if (!Arrays.equals(durable.txIds(), new long[]{1})) {
                        throw new IllegalStateException("unexpected reopened durable tx ids");
                    }
                    System.out.println("sqlite-reopened rows: " + rows.rowCount());
                    if (rows.rowCount() != 1) {
                        throw new IllegalStateException("unexpected SQLite reopened row count");
                    }
                }
            } finally {
                deleteSqliteFiles(sqlitePath);
            }
        } finally {
            vev.close();
        }
    }
}
